### PR TITLE
Kinematic Trajectory Optimization draft PR

### PR DESCRIPTION
**Pull request by kamiradi**
This PR aims to transcribe a basic Kinematic Trajectory Optimization problem from Moveit to Drake

- [ ]  Overload the PlannerManager class
- [ ]  Overload the PlanningContext class
- [ ]  Implement a ktopt_interface class

 Implement an example demo to test and integrate the above classes
PS: this is a draft, and the work may be split over a sequence of PRs

### CODE REVIEW

**Review by sea-bass — COMMENTED** [review]
Looks good! I made some of the nitpicky comments.

As far as when I think we're at a mergeable state here: Ensure we can not only set up the `KinematicTrajectoryOptimization` program, but that we can actually **solve** it. Doesn't need to have any constraints. And then return the outcome as a planning response.

While I waited an eternity for the Docker container to keep building without OOMing my machine, I filled in some more placeholder code. Hope that's okay.

File: .docker/Dockerfile
Comment (sea-bass) [inline .docker/Dockerfile:69]: @kamiradi I had to fix this because the paths were wrong, and the `--force` flag was then needed because this is trying to clone `moveit2`, `moveit_resources`, and `moveit_msgs` as duplicate repos.

I can't say I follow this very well, but just letting you know.

**Comment by sea-bass** [comment]
Made more updates to the Dockerfile. Now I can do:

* `docker compose build`
* `docker compose run moveit`
* `$ colcon build --packages-select moveit_drake`
* Run the demo

And there are still no `moveit` apt packages installed. Ever! No need to remove any extra ones either :)

**Review by kamiradi — COMMENTED** [review]
@sea-bass, I see that you have added the relevant bits to report the motion plan response. Were you able to test this on the demo?

**Comment by sea-bass** [comment]
> @sea-bass, I see that you have added the relevant bits to report the motion plan response. Were you able to test this on the demo?

I was, but it's far from complete. There are no actual costs/constraints added, and the Drake trajectory needs to be converted to a MoveIt `RobotTrajectory` ans packed up into the result.

File: src/ktopt_planning_context.cpp
Comment (sjahr) [inline src/ktopt_planning_context.cpp:47]: I guess this will work for now, but maybe this step needs to be dropped and we let ktopt handle the goal constraints directly

File: src/ktopt_planning_context.cpp
Comment (sjahr) [inline src/ktopt_planning_context.cpp:76]: I guess this could be exposed as a ROS parameter to let the use define the trajectory density

File: src/ktopt_planning_context.cpp
Comment (sjahr) [inline src/ktopt_planning_context.cpp:89]: Print a warning here that termination is not implemented by this planner

File: src/ktopt_planning_context.cpp
Comment (sjahr) [inline src/ktopt_planning_context.cpp:105]: I wonder if this (in a follow-up PR) should be moved into the constructor or even the planner manager

File: src/ktopt_planning_context.cpp
Comment (sjahr) [inline src/ktopt_planning_context.cpp:110]: Just print  warning that is it not implemented for now :wink:

**Review by sjahr — COMMENTED** [review]
Lots of small comments but in general this looks good! Once they are either addressed in the code or by converting the bigger ones into issues, we should merge this

File: demo/launch/pipeline_testbench.launch.py
Comment (kamiradi) [reply demo/launch/pipeline_testbench.launch.py:113]: Yeah, I just left everything as is from the STOMP implementation just in case, I'll do a pass of optimizations and remove everything that is not required

Comment (kamiradi) [reply demo/launch/pipeline_testbench.launch.py:113]: Yup, will do a pass of optimizations to remove unnecessary options. For now creating an issues

File: demo/launch/pipeline_testbench.launch.py
Comment (kamiradi) [reply demo/launch/pipeline_testbench.launch.py:122]: Creating an issue for future pass.

File: include/ktopt_interface/ktopt_planning_context.hpp
Comment (kamiradi) [reply include/ktopt_interface/ktopt_planning_context.hpp:56]: I am still trying to figure out how Drake declares these pointers as these need to match what the `AddMultibodySceneGraph` function returns. I tried a couple of options, this one was building. I'll go through this in more detail when I actually try and transript the planning scene into scene graph

Comment (kamiradi) [reply include/ktopt_interface/ktopt_planning_context.hpp:56]: I am still trying to figure out the right pointer declaration for the SceneGraph that matches what is returned by the `AddMultibodySceneGraph` function within drake. I tried a couple, this was the one that was building. I'll inspect further once I get to transcribing the planning scene in a more principled way. For now creating an issue.

File: src/ktopt_planner_manager.cpp
Comment (kamiradi) [reply src/ktopt_planner_manager.cpp:83]: Yeah, I was planning to add a warning message around this for now.

Comment (kamiradi) [reply src/ktopt_planner_manager.cpp:83]: Yeah, I was planning on adding a wanring message for this.

File: src/ktopt_planner_manager.cpp
Comment (kamiradi) [reply src/ktopt_planner_manager.cpp:106]: yeah I thought about this, for now I did it this way as the ros handle is passed into planner_manager. I could alternatively declare a ros event handler within planning _context. But thought i'll defer this decision to later.

File: src/ktopt_planning_context.cpp
Comment (kamiradi) [reply src/ktopt_planning_context.cpp:76]: Completely agreed! That was going to be the next step.

File: src/ktopt_planning_context.cpp
Comment (kamiradi) [reply src/ktopt_planning_context.cpp:105]: Again, I declared this as a function because , there could be use in resetting the robot_description once a planning_context has been set. The alternative is to expose a ros_handler and pass it to the planner_manager. I'll experiment with that in a future pass.

File: README.md
Comment (sea-bass) [inline README.md:87]: Typo: `moveit_drake`

File: CMakeLists.txt
Comment (kamiradi) [reply CMakeLists.txt:54]: Is line 56 with the double ktopt_interface line right?

Comment (kamiradi) [reply CMakeLists.txt:54]: ot maybe a typo?

File: include/ktopt_interface/ktopt_planning_context.hpp
Comment (sjahr) [reply include/ktopt_interface/ktopt_planning_context.hpp:28]: Yes

**Review by kamiradi — COMMENTED** [review]
So I have made a few changes to the repository based on your comments and create a few follow on issues to be tracked with later commits.

File: CMakeLists.txt
Comment (sjahr) [reply CMakeLists.txt:54]: Oh yes :+1:

**Comment by sjahr** [comment]
@kamiradi Can you ping when you addressed all comments for another review round? You can resolve the conversation you addressed and comment on the ones where you might be unsure/ have questions or commetns

File: include/ktopt_interface/ktopt_planning_context.hpp
Comment (kamiradi) [reply include/ktopt_interface/ktopt_planning_context.hpp:28]: creating issue.

File: res/drake_moveit.yaml
Comment (kamiradi) [reply res/drake_moveit.yaml:4]: Creating issue.

Comment (kamiradi) [reply res/drake_moveit.yaml:4]: creating issue.

File: src/ktopt_planner_manager.cpp
Comment (kamiradi) [reply src/ktopt_planner_manager.cpp:36]: created issue.

File: src/ktopt_planner_manager.cpp
Comment (kamiradi) [reply src/ktopt_planner_manager.cpp]: changed in commit.

**Comment by kamiradi** [comment]
@sjahr, I remember you pointing me towards a specific yaml file that needs to have a `_planning` suffix to be loaded by the class loader. Could you point me towards an example? I believe this is needed for the drake planning plugin to be loaded during runtime.

File: src/ktopt_planning_context.cpp
Comment (kamiradi) [reply src/ktopt_planning_context.cpp]: creating issue.

**Comment by sjahr** [comment]
> @sjahr, I remember you pointing me towards a specific yaml file that needs to have a `_planning` suffix to be loaded by the class loader. Could you point me towards an example? I believe this is needed for the drake planning plugin to be loaded during runtime.
@kamiradi The file is specified per moveit config, for example in panda_moveit_config for stomp: 
https://github.com/moveit/moveit_resources/blob/ros2/panda_moveit_config/config/stomp_planning.yaml

The content is equivalent with the section you've added to the launch file and either the yaml file or the launch file section can be used (although the .yaml file is best practices for fleshed out plugins).

**Comment by kamiradi** [comment]
Ah I see, so this is an alternative to the launch file config mentioned in the launch file. Is there a runtime requirement for the yaml file also to be present? Can I defer this to the next PR?

So the requirement now is to:

1. Declare a similar `ktopt_planning.yaml` file (where should I be doing this? should I create a new config folder within the repository?)
2. And the class loader automagically loads this at runtime.

File: demo/launch/pipeline_testbench.launch.py
Comment (kamiradi) [inline demo/launch/pipeline_testbench.launch.py:1]: drake moveit pipeline test

Comment (sea-bass) [reply demo/launch/pipeline_testbench.launch.py:1]: I found a typo and fixed in https://github.com/moveit/moveit_drake/pull/8/commits/2ff0c3b67aec4214d92ff01788a9c9297acdf356 -- try with that?

**Review by sea-bass — APPROVED** [review]
Approving pending @sjahr's final review.

My only suggestion: Rename the PR title to something like "Kinematic Trajectory Optimization skeleton"

File: CMakeLists.txt
Comment (sjahr) [inline CMakeLists.txt:52]: @kamiradi @sea-bass I needed to set these properties to avoid a segfault when the plugin was loaded because it couldn't find libdrake at runtime. Let me know what you think

**Review by sjahr — APPROVED** [review]
Works for me now :tada:

**Pull request merged**
